Fix/rails autoloading deprecations

ulferts opened the pull request:
Attempts to get rid of referencing autoloaded objects in initializers which would lead to having multiple incarnations of the same object when unloading. Rails warns about those cases as the behaviour will no longer be supported in rails 7:

```
DEPRECATION WARNING: Initialization autoloaded the constants JobStatus, JobStatus::ApplicationJobWithStatus, OpenProject::Database, API, API::V3, API::V3::Utilities, API::Utilities, OpenProject::StaticRouting, API::Utilities::UrlHelper, API::V3::Utilities::PathHelper, OpenProjectErrorHelper, FrontendAssetHelper, OpenProject::Logging::SentryLogger, ApplicationRecord, OpenProject::Deprecation, Setting::CallbacksHelper, Setting::Aliases, Setting, Sessions, Sessions::SqlBypass, OpenProject::Notifications, OpenProject::Events, UserInvitation, OpenProject::Authentication::Strategies::Warden::AnonymousFallback, AngularHelper, and OpenProject::JobStatus::EventListener.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload JobStatus, for example,
the expected changes won't be reflected in that stale Module object.

These autoloaded constants have been unloaded.

In order to autoload safely at boot time, please wrap your code in a reloader
callback this way:

    Rails.application.reloader.to_prepare do
      # Autoload classes and modules needed at boot time here.
    end

That block runs when the application boots, and every time there is a reload.
For historical reasons, it may run twice, so it has to be idempotent.

Check the "Autoloading and Reloading Constants" guide to learn more about how
Rails autoloads and reloads.
```

## Patterns

### Patches

Patches come in two flavors: Patches to the Rails framework and engines patching OpenProject core. For both, the same mechanism on when to load the code can be taken which is inspired by [a rails guid on engines](https://guides.rubyonrails.org/engines.html#load-and-configuration-hooks). It would require us to add calls to 
```
ActiveSupport.run_load_hooks(:some_key, SomeClass)
```
to files that need patching.
For a couple of rails framework files, those calls already exist. 

Patches that need to happen during initialization, will have to avoid being unloaded.

### Configuration

OpenProject has a lot of Objects that can be configured (e.g. Menu, Permissions, Activity) in other parts of the application be that core or module. Nowadays this oftentimes happens in initializers which isn't ideal. 
With load_hooks in place (see 'Patches'), the configuration can happen after the object is referenced for the first time.  

## TODO

* [x] Either use hooks, wrap in `to_prepare` or place outside of an autoloaded folder:
  - [x] Redmine::I18n
  - [x] OpenProject::NullDbFallback
  - [x] Redmine::Diff
  - [x] Redmine::Diff::Diffable
  - [x] JobStatus
  - [x] JobStatus::ApplicationJobWithStatus
  - [x] OpenProject::TextFormatting::Truncation
  - [x] OpenProject::TextFormatting
  - [x] OpenProject::ObjectLinking
  - [x] OpenProject::SafeParams
  - [x] HookHelper
  - [x] IconsHelper
  - [x] AuthenticationStagePathHelper
  - [x] AdditionalUrlHelpers
  - [x] OpenProject::PageHierarchyHelper
  - [x] ApplicationHelper
  - [x] OpenProject::Database
  - [x] API
  - [x] API::V3
  - [x] API::V3::Utilities
  - [x] API::Utilities
  - [x] OpenProject::StaticRouting
  - [x] API::Utilities::UrlHelper
  - [x] API::V3::Utilities::PathHelper
  - [x] OpenProject::AccessControl::Mapper
  - [x] OpenProject::AccessControl::Permission
  - [x] OpenProjectErrorHelper
  - [x] FrontendAssetHelper
  - [x] OpenProject::Logging::SentryLogger
  - [x] ApplicationRecord
  - [x] OpenProject::Deprecation
  - [x] Setting::CallbacksHelper
  - [x] Setting::Aliases
  - [x] Setting
  - [x] Sessions
  - [x] Sessions::SqlBypass
  - [x] OpenProject::Notifications
  - [x] OpenProject::Events
  - [x] UserInvitation
  - [x] OpenProject::Authentication::Strategies::Warden::AnonymousFallback
  - [x] RbCommonHelper
  - [x] AngularHelper
  - [x] RailsCell
  - [x] OpenProject::JobStatus::EventListener

* [ ] Cleanup as above may have been done in a quick&dirty fashion 
  * [ ] Find a suitable place for non autoloaded and non unloaded objects (`lib` is autoloaded currently and a subdirectory therein cannot be used)
  * [ ] Move contents of `config/constants` to new place
  * [ ] provide helpers in `acts_as_op_engine` to avoid changes in plugins and having to repeat `to_prepare` within a plugin.

ulferts committed:
move null_db_fallback to constants

ulferts committed:
avoid reference to Project in initializer

ulferts committed:
avoid autoloading ApplicationHelper during initialization

ulferts committed:
defer autoloading permissions

ulferts committed:
use active_support hook for patching

ulferts committed:
introduce lib_static

ulferts committed:
defer configuring session store

ulferts committed:
defer registering for op notification

ulferts committed:
turn OpenProject::Notifications static

ulferts committed:
static op/authentication

ulferts committed:
turn op/database static

oliverguenther committed:
Put status job listener in prepare block

oliverguenther committed:
Load patches in autoloader

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into fix/rails_autoloading_deprecations

oliverguenther committed:
Move grape initializer into reloader

oliverguenther committed:
Move bim engine registration in prepare

oliverguenther committed:
Move rack_timeout in reloader

oliverguenther committed:
Move interceptors registry into reloader

oliverguenther committed:
Move secure_headers into reloader

oliverguenther committed:
Move sentry into reloader

oliverguenther committed:
Linting

oliverguenther commented:
For some reason, c794a9249bcef1422d60991fb5e81821e64f82a9 is breaking a large number of tests due to changes in the way project modules behave. I haven't yet gotten behind it.

/edit, found at least one reason: https://github.com/opf/openproject/commit/c794a9249bcef1422d60991fb5e81821e64f82a9#diff-bfd7411e8846ca3e7270ea392634fffc5b11f8bf44b07549c2f33857d65fcfbdL333 

`@project_scope` is of course unset on reloading, so this is no longer applying permissions to a project scope. The whole way of dealing with this variable is painful though, maybe we can get rid of it right away.

oliverguenther committed:
Clear caches when mapping new modules in access control

oliverguenther committed:
Fix unsetting the @project_scope instance variable

on reload, this variable obviously is no longer present

oliverguenther committed:
Move configuration to static lib

oliverguenther committed:
Remove legacy mailer configuration and move into settings

Split from "Remove Setting.protocol in favor of static config"

Restore cache store setup

oliverguenther committed:
Remove Setting.protocol in favor of static config

oliverguenther committed:
Replace to_prepare with after_initialize for one-time calls

oliverguenther committed:
Remove bcrypt override deprecation
